File: src/db/crud/events.py
from typing import Any, Dict, List, Optional

from src.db.d1_client import d1_exec, d1_one, d1_query


def get_preferences(owner_id: int) -> Optional[Dict[str, Any]]:
    row = d1_one("SELECT * FROM user_event_preferences WHERE owner_id = ?", [owner_id])
    return dict(row) if row else None


def set_preferences(owner_id: int, city: Optional[str], radius_km: Optional[int]) -> Dict[str, Any]:
    existing = get_preferences(owner_id)
    if existing:
        d1_exec(
            "UPDATE user_event_preferences SET city = ?, radius_km = COALESCE(?, radius_km) WHERE owner_id = ?",
            [city, radius_km, owner_id],
        )
    else:
        d1_exec(
            "INSERT INTO user_event_preferences (owner_id, city, radius_km) VALUES (?, ?, COALESCE(?, 30))",
            [owner_id, city, radius_km],
        )
    return get_preferences(owner_id)


def list_events(city: Optional[str] = None, upcoming_only: bool = True) -> List[Dict[str, Any]]:
    where: List[str] = []
    params: List[Any] = []
    if city:
        where.append("city = ?")
        params.append(city)
    if upcoming_only:
        where.append("date(event_date) >= date('now')")

    clause = f"WHERE {' AND '.join(where)}" if where else ""
    rows = d1_query(f"SELECT * FROM literary_events {clause} ORDER BY event_date ASC", params)
    return [dict(r) for r in rows]


def get_event(event_id: int) -> Optional[Dict[str, Any]]:
    row = d1_one("SELECT * FROM literary_events WHERE event_id = ?", [event_id])
    return dict(row) if row else None


def create_manual_event(
    title: str,
    event_date: str,
    description: Optional[str] = None,
    venue_name: Optional[str] = None,
    city: Optional[str] = None,
    url: Optional[str] = None,
    image_url: Optional[str] = None,
) -> Dict[str, Any]:
    meta = d1_exec(
        """
        INSERT INTO literary_events (source, title, description, venue_name, city, event_date, url, image_url)
        VALUES ('manual', ?, ?, ?, ?, ?, ?, ?)
        """,
        [title, description, venue_name, city, event_date, url, image_url],
    )
    return get_event(meta["last_row_id"])


def delete_manual_event(event_id: int) -> None:
    """Only 'manual' rows are deletable — provider-sourced rows are refreshed by the weekly job, not user-managed."""
    d1_exec("DELETE FROM literary_events WHERE event_id = ? AND source = 'manual'", [event_id])

File: cloudflare/src/events.ts
import type { Env } from "./index";
import { insertNotificationsIfMissing, type NotificationCandidate } from "./notifications";

interface DiscoveredEvent {
  external_id: string;
  title: string;
  description: string | null;
  venue_name: string | null;
  city: string;
  event_date: string;
  url: string | null;
  image_url: string | null;
}

interface TicketmasterResponse {
  _embedded?: {
    events?: {
      id: string;
      name: string;
      info?: string;
      pleaseNote?: string;
      url?: string;
      images?: { url: string }[];
      dates?: { start?: { localDate?: string } };
      _embedded?: { venues?: { name?: string }[] };
    }[];
  };
}

// Interface, not a hardcoded call (section 4.5): Ticketmaster is the only
// source implemented today (free dev key, real search API — unlike
// Eventbrite, whose public discovery API was discontinued). SerpApi (paid)
// could be added as a second provider later without touching the job
// below. Manual entries (source: 'manual') never go through a provider —
// they're written directly by the /events POST endpoint as the
// always-available fallback for coverage a search API won't have.
type EventProvider = (city: string, env: Env) => Promise<DiscoveredEvent[]>;

const LITERARY_KEYWORDS = "book fair OR literary festival OR author signing OR book launch";

async function fetchTicketmasterEvents(city: string, env: Env): Promise<DiscoveredEvent[]> {
  if (!env.TICKETMASTER_API_KEY) return [];

  const params = new URLSearchParams({
    apikey: env.TICKETMASTER_API_KEY,
    city,
    keyword: LITERARY_KEYWORDS,
    sort: "date,asc",
  });
  const res = await fetch(`https://app.ticketmaster.com/discovery/v2/events.json?${params}`);
  if (!res.ok) return [];

  const data = await res.json<TicketmasterResponse>();
  const events = data._embedded?.events ?? [];

  return events
    .filter((e) => e.dates?.start?.localDate)
    .map((e) => ({
      external_id: e.id,
      title: e.name,
      description: e.info ?? e.pleaseNote ?? null,
      venue_name: e._embedded?.venues?.[0]?.name ?? null,
      city,
      event_date: e.dates!.start!.localDate!,
      url: e.url ?? null,
      image_url: e.images?.[0]?.url ?? null,
    }));
}

const EVENT_PROVIDERS: EventProvider[] = [fetchTicketmasterEvents];

async function getWatchedCities(env: Env): Promise<Map<string, number[]>> {
  const result = await env.DB.prepare(
    `SELECT owner_id, city FROM user_event_preferences WHERE city IS NOT NULL AND trim(city) != ''`
  ).all<{ owner_id: number; city: string }>();

  // Group owners by city so a city watched by multiple owners is only
  // queried against the provider once per run.
  const byCity = new Map<string, number[]>();
  for (const row of result.results) {
    const key = row.city.trim();
    if (!byCity.has(key)) byCity.set(key, []);
    byCity.get(key)!.push(row.owner_id);
  }
  return byCity;
}

// Upserts on (source, external_id) so re-running the weekly job doesn't
// duplicate rows or re-notify for an event already seen — only a genuinely
// new external_id produces a notification.
async function upsertEvent(env: Env, ev: DiscoveredEvent): Promise<{ eventId: number; isNew: boolean }> {
  const existing = await env.DB.prepare(
    `SELECT event_id FROM literary_events WHERE source = 'ticketmaster' AND external_id = ?`
  )
    .bind(ev.external_id)
    .first<{ event_id: number }>();

  if (existing) {
    await env.DB.prepare(
      `UPDATE literary_events SET title=?, description=?, venue_name=?, city=?, event_date=?, url=?, image_url=?, cached_at=datetime('now') WHERE event_id=?`
    )
      .bind(ev.title, ev.description, ev.venue_name, ev.city, ev.event_date, ev.url, ev.image_url, existing.event_id)
      .run();
    return { eventId: existing.event_id, isNew: false };
  }

  const inserted = await env.DB.prepare(
    `INSERT INTO literary_events (source, external_id, title, description, venue_name, city, event_date, url, image_url)
     VALUES ('ticketmaster', ?, ?, ?, ?, ?, ?, ?, ?)`
  )
    .bind(ev.external_id, ev.title, ev.description, ev.venue_name, ev.city, ev.event_date, ev.url, ev.image_url)
    .run();
  return { eventId: Number(inserted.meta.last_row_id), isNew: true };
}

export async function checkLiteraryEventsAndNotify(
  env: Env
): Promise<{ citiesChecked: number; eventsUpserted: number; inAppCreated: number }> {
  const ownersByCity = await getWatchedCities(env);
  const candidates: NotificationCandidate[] = [];
  let eventsUpserted = 0;

  for (const [city, owners] of ownersByCity) {
    let events: DiscoveredEvent[] = [];
    for (const provider of EVENT_PROVIDERS) {
      events = await provider(city, env);
      if (events.length > 0) break;
    }
    if (events.length === 0) continue; // no provider had results — skip, not an error

    for (const ev of events) {
      const { isNew } = await upsertEvent(env, ev);
      eventsUpserted++;
      if (!isNew) continue;

      for (const ownerId of owners) {
        candidates.push({
          owner_id: ownerId,
          type: "EVENT_NEARBY",
          title: `📚 ${ev.title}`,
          body: `${ev.venue_name ? `${ev.venue_name} — ` : ""}${ev.event_date}`,
          link: "/events",
        });
      }
    }
  }

  const inAppCreated = await insertNotificationsIfMissing(env, candidates);
  return { citiesChecked: ownersByCity.size, eventsUpserted, inAppCreated };
}
